Fixes #25113 - Refactor Org handling in React

johnpmitsch opened the pull request:
This commit refactors org handling in our React pages by updating the `withOrganization` component to handle the Organization logic and then passing the Organization to the wrapped component. This keeps the logic separate and always tracks the Organization in Foreman.

My understanding of how Foreman handles an organization is this:

- Foreman passes the current org title from [the rails side](https://github.com/theforeman/foreman/blob/84ae69c6bf191837e8c7b4ccadb4583eebeda172/app/helpers/layout_helper.rb#L39) to react when the components [are mounted](https://github.com/theforeman/foreman/blob/84ae69c6bf191837e8c7b4ccadb4583eebeda172/app/helpers/layout_helper.rb#L3)
- A currentOrganization prop is passed to Layout from Redux.
  - The layout reducer [starts with "Any Organization"](https://github.com/theforeman/foreman/blob/84ae69c6bf191837e8c7b4ccadb4583eebeda172/webpack/assets/javascripts/react_app/components/Layout/LayoutReducer.js#L21)
  - The [`changeOrganization`](https://github.com/theforeman/foreman/blob/84ae69c6bf191837e8c7b4ccadb4583eebeda172/webpack/assets/javascripts/react_app/components/Layout/LayoutActions.js#L40-L45) action can change this value in Redux
- Foreman checks if the `currentOrganization` is the same as what is being passed in from Rails [here](https://github.com/theforeman/foreman/blob/84ae69c6bf191837e8c7b4ccadb4583eebeda172/webpack/assets/javascripts/react_app/components/Layout/Layout.js#L49-L59) and updating if they are not the same.
- So as long as the `Layout` component is rendering, the Org in Redux always equals the org set in Rails

Then in Katello:

- The index file for the top-level `Application` Component [loads the organization](https://github.com/Katello/katello/blob/master/webpack/containers/Application/index.js#L22-L25)
  - This fetches the organization details [with an action](https://github.com/Katello/katello/blob/master/webpack/scenes/Organizations/OrganizationActions.js#L14-L35) and uses the [`orgId()` function](https://github.com/Katello/katello/blob/master/webpack/services/api/index.js#L116-L122) which pulls the organization from a data attribute on the html element itself
  - Then it is [saved by the reducer](https://github.com/Katello/katello/blob/master/webpack/scenes/Organizations/OrganizationReducer.js#L22)
  - This is then passed back to the Application as a prop [here](https://github.com/Katello/katello/blob/6ef62d1f497f905848f6185a279ca12d0c170eb8/webpack/containers/Application/index.js#L12)

(I had to write this out for my own understanding too)

This means we can use the `layout` organization from Foreman to track the organization in Katello instead of grabbing the org id from the DOM itself. We also have more information added to the org in Katello, so we need to make the call to load the org using our API. This logic is moved to the `withOrganization` HOC.

The `withOrganization` logic is:

- Is the Foreman org "Any Organization?"
  - Show org switcher
- else is the Foreman Org loaded?
  - Load the org in Katello if it is not loaded or doesn't match
  - Once the org is loaded into Katello's store, render the wrapped component and pass the org

The goal is to always keep the Foreman and Katello organizations in sync and render the wrapped components when they are in sync.

This commit also adds a redirect back to the base page when a user is on a subpage and switches organizations. For example, I am on `subscriptions/34` and I change my org in the org switcher, I will be redirected back to `/subscriptions`. 

There are also updated tests to reflect this new logic

johnpmitsch commented:
> For me there is a little bit of a delay when switching from one org to another. It shows the organization details page for a second before switching to the subscriptions list page. To me this is acceptable, however, as it's a vast improvement over the old version.

Ya, I noticed the same thing and had the same thought. Unfortunately, it takes that time to get to the redirect as the organization is being fetched. (at least this is my current understanding of things)

There is a weird issue where the org selector stays on `Any Organization` after doing "Flow 2" in the associated redmine bug. I'm looking into it, its causing some issues  if you continue to do anything on that page without refreshing.

waldenraines commented:
> There is a weird issue where the org selector stays on `Any Organization` after doing "Flow 2" in the associated redmine bug. I'm looking into it, its causing some issues if you continue to do anything on that page without refreshing.

Yeah, that was the exact reason I opted for the page refresh.  Hope you're able to figure out a better solution.

waldenraines commented:
> Yeah, that was the exact reason I opted for the page refresh. Hope you're able to figure out a better solution.

@johnpmitsch any updates on this?

johnpmitsch commented:
> @johnpmitsch any updates on this?

None yet, but I haven't had a good chance to look. I'll take a look soon and update the PR, thanks for the reminder poke.

johnpmitsch commented:
I got farther with this, the org is now switching correctly after the `Any Organization` workflow.

However, there is an error when the page  `Organization with id undefined not found` after switching. This is from the manifest history API call, where `orgId()` is returning undefined. It looks like we grab that from the DOM by way of loading it in the ERB template and it is blank with my recent changes. I'll keep digging!

johnpmitsch commented:
I pushed up what I have so far in case anyone is curious.

waldenraines commented on webpack/components/SelectOrg/SelectOrgReducer.js:
```diff
@@ -19,11 +19,13 @@ export default (state = initialState, action) => {
-    case CHANGE_CURRENT_ORGANIZATION_SUCCESS:
+    case CHANGE_CURRENT_ORGANIZATION_SUCCESS: {
+      const { id, title } = payload;
+      window.tfm.nav.changeOrganization({id: id, title: title});
```
I am seeing this causing this error:

```
  Error: Reducers may not dispatch actions.
    at dispatch (createStore.js:160)
    at index.js:104
    at index.js:11
    at dispatch (applyMiddleware.js:35)
    at LayoutActions.js:46
    at Object.dispatch (index.js:8)
    at Object.changeOrganization (foreman_navigation.js:17)
    at exports.default (SelectOrgReducer.js:24)
    at combination (combineReducers.js:120)
    at combination (combineReducers.js:120)
```

johnpmitsch commented:
@waldenraines Thanks for sharing that, you are right that is better in the redux action. In the `Any organization` workflow,  It doesn't look like the org will update in the org switcher, though I am seeing the correct subs loading.

I'm continuing to look and also looking at foreman and I'm getting closer. I think some improvements in how we are handling the org will help not only this bug but prevent issues like this later on.

johnpmitsch commented:
@waldenraines now that I have more time to work on this, I've looked further and am getting closer. the workflows are mostly fixed but I'm still able to get things into a weird 'mismatched' state if I keep switching around orgs and pages. 

I'm continuing to dig and understand the chain of events better and I may have some ideas around larger refactoring to make our org handling more stable.

For now, I've pushed up what I have. Thanks for the patience on this one :)

waldenraines commented:
Thanks @johnpmitsch.  Perhaps @amirfefer can assist you if you have any questions about the organization handling.

johnpmitsch commented:
@waldenraines @amirfefer I've gotten this working and I think it will be a cleaner way to handle the org in Katello and pass the org info down to the child components. I'll do a more comprehensive write up about the changes and switch from a Draft PR when I get the tests passing and address the following issue(s). For now, feel free to review and comment on the PR about the overall direction of things

Aside from tests and cleanup, there is one major issue and one small issue to address:

1) (the big issue) when I go to a subscription details page (i.e. subscriptions/44) and then select "Any Organization", sometimes the page will go blank and I'll be redirected to `organizations/clear`. It looks like that is explicitly done [here in foreman's component](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L32), I'm guessing it is necessary to clear the current organization, but it seems using it as an `href` actually redirects the browser there. This being said, I'm not exactly sure why its an issue only on subpages, I'm guessing the page is redirected back to the `redirectPage` that is passed in to `withOrganization` and the `href` from `TaxonomyDropdown` is also redirecting but it happens after the `redirectPage`. Any insight into this would be appreciated! (P.S. for the question "why haven't we seen this before?", this workflow is one of the broken ones in the original issue for this PR, so fixing the workflow exposed this issue.)

2) (a smaller issue) When selecting `Any Organization`, the SetOrganization page loads twice. You get a small flash of the page, and then it reloads again. It's minor, but the component is mounting twice. I think this is [actually an issue with HOCs and react-router](https://stackoverflow.com/a/49056652). I don't see it being a blocking issue but want to make sure I call it out :)

johnpmitsch commented:
Another thing I noticed that is tangentially related to this PR is [calling setTimeout in LoadingState](https://github.com/Katello/katello/blob/master/webpack/move_to_pf/LoadingState/LoadingState.js#L15-L17) is causing warnings in the console. 

I'm seeing a lot of `Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.`

This is because its an asynchronous task, so the component is unmounted, but the state is then updated after the timeout. It's not affecting anything, but with this PR, the warnings become more frequent as `LoadingState` is now in `withOrg`, which is used by every page.

@amirfefer perhaps there is a better way to set a delay for the page loading spinner?

johnpmitsch commented:
I also see a lot of warnings on the module stream page, since I see it in master too, I filed this: https://projects.theforeman.org/issues/26861

johnpmitsch commented:
@waldenraines @amirfefer I updated this and moved it away from a draft PR, but I do still have the first issue [I mention here](https://github.com/Katello/katello/pull/8052#issuecomment-494955513) to resolve

johnpmitsch commented:
> 1. (the big issue) when I go to a subscription details page (i.e. subscriptions/44) and then select "Any Organization", sometimes the page will go blank and I'll be redirected to `organizations/clear`. It looks like that is explicitly done [here in foreman's component](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L32), I'm guessing it is necessary to clear the current organization, but it seems using it as an `href` actually redirects the browser there. This being said, I'm not exactly sure why its an issue only on subpages, I'm guessing the page is redirected back to the `redirectPage` that is passed in to `withOrganization` and the `href` from `TaxonomyDropdown` is also redirecting but it happens after the `redirectPage`. Any insight into this would be appreciated! (P.S. for the question "why haven't we seen this before?", this workflow is one of the broken ones in the original issue for this PR, so fixing the workflow potentially exposed this issue.)

@glekner do you have any insight into this issue and how we can handle it here?

waldenraines commented:
> 1. (the big issue) when I go to a subscription details page (i.e. subscriptions/44) and then select "Any Organization", sometimes the page will go blank and I'll be redirected to `organizations/clear`. It looks like that is explicitly done [here in foreman's component](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L32), I'm guessing it is necessary to clear the current organization, but it seems using it as an `href` actually redirects the browser there. This being said, I'm not exactly sure why its an issue only on subpages, I'm guessing the page is redirected back to the `redirectPage` that is passed in to `withOrganization` and the `href` from `TaxonomyDropdown` is also redirecting but it happens after the `redirectPage`. Any insight into this would be appreciated! (P.S. for the question "why haven't we seen this before?", this workflow is one of the broken ones in the original issue for this PR, so fixing the workflow potentially exposed this issue.)

@johnpmitsch do you think that we could fix this in a future PR or do you think it's needed for this one?

glekner commented:
> > 1. (the big issue) when I go to a subscription details page (i.e. subscriptions/44) and then select "Any Organization", sometimes the page will go blank and I'll be redirected to `organizations/clear`. It looks like that is explicitly done [here in foreman's component](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L32), I'm guessing it is necessary to clear the current organization, but it seems using it as an `href` actually redirects the browser there. This being said, I'm not exactly sure why its an issue only on subpages, I'm guessing the page is redirected back to the `redirectPage` that is passed in to `withOrganization` and the `href` from `TaxonomyDropdown` is also redirecting but it happens after the `redirectPage`. Any insight into this would be appreciated! (P.S. for the question "why haven't we seen this before?", this workflow is one of the broken ones in the original issue for this PR, so fixing the workflow potentially exposed this issue.)
> 
> @glekner do you have any insight into this issue and how we can handle it here?

IIRC we used an `href` so the server can clear the taxonomy aswell. which routing tech is being used in the sub page? (angular?). While i think about it, do we really need a redirect here to occur? maybe we could just trigger an API call to update the server.

johnpmitsch commented:
> > > 1. (the big issue) when I go to a subscription details page (i.e. subscriptions/44) and then select "Any Organization", sometimes the page will go blank and I'll be redirected to `organizations/clear`. It looks like that is explicitly done [here in foreman's component](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L32), I'm guessing it is necessary to clear the current organization, but it seems using it as an `href` actually redirects the browser there. This being said, I'm not exactly sure why its an issue only on subpages, I'm guessing the page is redirected back to the `redirectPage` that is passed in to `withOrganization` and the `href` from `TaxonomyDropdown` is also redirecting but it happens after the `redirectPage`. Any insight into this would be appreciated! (P.S. for the question "why haven't we seen this before?", this workflow is one of the broken ones in the original issue for this PR, so fixing the workflow potentially exposed this issue.)
> > 
> > 
> > @glekner do you have any insight into this issue and how we can handle it here?
> 
> IIRC we used an `href` so the server can clear the taxonomy aswell. which routing tech is being used in the sub page? (angular?). While i think about it, do we really need a redirect here to occur? maybe we could just trigger an API call to update the server.

@glekner This is what I was thinking as well, I think the href will be hard to work around and we should do an API call instead.



> @johnpmitsch do you think that we could fix this in a future PR or do you think it's needed for this one?

@waldenraines yes, I'm fine with that, especially if we go with the above proposed solution.

waldenraines commented:
> @glekner This is what I was thinking as well, I think the href will be hard to work around and we should do an API call instead.

:+1: 

From what I can tell there is currently no API that supports this action.  Since this fix is for an upcoming release I wonder if we could work on getting this in and then fix the clear action with a follow up PR.

jturel commented:
@johnpmitsch this is working nicely and addresses the reported issue - I also don't see any problems with the code itself.

One thing I did notice is when attempting to reproduce the reported problem I occasionally get this stack trace once redirected back to /subscriptions:

```
react-dom.development.js:506 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in Search (created by SubscriptionsToolbar)
    in div (created by FormGroup)
    in FormGroup (created by SubscriptionsToolbar)
    in form (created by Form)
    in Form (created by SubscriptionsToolbar)
    in div (created by Col)
    in Col (created by SubscriptionsToolbar)
    in div (created by Row)
    in Row (created by SubscriptionsToolbar)
    in SubscriptionsToolbar (created by SubscriptionsPage)
    in div (created by Col)
    in Col (created by SubscriptionsPage)
    in div (created by Row)
    in Row (created by SubscriptionsPage)
    in div (created by Grid)
    in Grid (created by SubscriptionsPage)
    in SubscriptionsPage (created by Route)
    in Route (created by withRouter(SubscriptionsPage))
    in withRouter(SubscriptionsPage) (created by Connect(withRouter(SubscriptionsPage)))
    in Connect(withRouter(SubscriptionsPage)) (created by Headers)
    in Headers (created by CheckOrg)
    in LoadingState (created by CheckOrg)
    in CheckOrg (created by Connect(CheckOrg))
    in Connect(CheckOrg) (created by Route)
    in Route
    in div
    in Unknown (created by Application)
```

It seems to happen maybe 1/4 of the time. An above comment led me to believe this was fixed, so I just wanted to mention it. Any ideas?

johnpmitsch commented:
> It seems to happen maybe 1/4 of the time. An above comment led me to believe this was fixed, so I just wanted to mention it. Any ideas?

@jturel Thanks for looking! Regarding the error, I did see that as well. It's actually the same error as [this comment](https://github.com/Katello/katello/pull/8052#issuecomment-494957522) but coming from a different component. This wasn't obvious since I didn't include a traceback, the one from my comment was coming from `LoadingState` and the one you mention is coming from `SubscriptionToolbar`. I don't see any changes made in that area in this PR. Could you check if the error you mention is present on master? I can look as well.

jturel commented:
I wasn't able to repro the stack trace I showed on master (not saying it's not possible to do so...) but after switching back to this code it seems that it's definitely less than 1/4 of the time and I guess it's some kind of race condition since it's unpredictable. I also suspect it's unrelated to this change since you didn't touch anything relating to the toolbar & search field as you mentioned. Let's keep an eye on it, though! :eyes:

johnpmitsch commented:
> From what I can tell there is currently no API that supports this action. Since this fix is for an upcoming release I wonder if we could work on getting this in and then fix the clear action with a follow up PR.

@waldenraines @glekner sure thing, I do wonder how we are going to handle it given there is no API, but I'm sure we can figure something out. 

If you could give this another look and comment or approve, we can get this moving along! :steam_locomotive:

waldenraines commented:
@sharvit @amirfefer @Laviro can one of you please have a look at this as well to make sure I didn't miss anything?

johnpmitsch commented:
I filed the blank page because of the redirect issue here: https://projects.theforeman.org/issues/27064

amirfefer commented:
Thanks for doing this refactoring @johnpmitsch ! :+1:   I'm reviewing it at this moment

sharvit commented:
> Thanks @johnpmitsch !
> There are two main issues about this refactoring:
> 
> 1. Mixing between inner state and redux - as I mentioned in the inline comment, we should avoid this in any cost, maybe it easier to write it at first but for long term this can cause pain and frustration.

+1

> 2. foreman already manages organization events in redux, therefore IMHO this responsibility should be manged in foreman core via redux,  Mutating the org from multiple mechanisms is risky, and it would be great to have one interface that responsible for it, that can be extend in plugins

+1

johnpmitsch commented:
> 2\. foreman already manages organization events in redux, therefore IMHO this responsibility should be manged in foreman core via redux,  Mutating the org from multiple mechanisms is risky, and it would be great to have one interface that responsible for it, that can be extend in plugins

@amirfefer @sharvit @waldenraines I should have touched on this earlier as its one of the parts that makes managing the org in a Katello difficult and the reason for some of the anti-patterns you seen here. Katello has its own API call for organization with the additional info organization provides. We are currently (in master) using this call to [load the organization in Katello](https://github.com/Katello/katello/blob/master/webpack/containers/Application/index.js#L22-L25). This uses `orgId()`. `orgId` tracks the current organization [by pulling it from the DOM node](https://github.com/Katello/katello/blob/master/webpack/services/api/index.js#L116-L122)

So my goal in this PR is to use the Foreman organization stored in Redux as that source of truth, and no longer use the DOM node as the source for the organization. To do this we must load the Katello organization when the Organization in Foreman changes. This is why you see me jumping through hoops in the `withOrg` component to do so using `getDerivedStateFromProps`. The Foreman org changes, Katello makes an API call, and then the wrapped component re-renders when the org is succesfully returned.

If we continue needing the org to load using Katello's API and keeping it in sync with Foreman's org, its going to be hard to do without using some hacky state management. We need to constantly "sync" up the org in Katello with Foreman's. At least, this is the conclusion I came to in my digging, I am open to any suggestions if we continue this route.

However, looking at things again, I'm not convinced we even need to load the organization in Katello? We add the `label` attribute to organization in Katello, along with [some other fields](https://github.com/Katello/katello/blob/master/db/migrate/20131120225132_add_organization_fields.rb#L3-L8)

If we just need the organization id and name, we could probably do just fine using Foreman's organization and not even using Katello's. If we do this, I think things could be simplified quite a bit. I don't know if there are historical reasons we load Katello's organization now. I would love to hear any more insight in this area and your thoughts of whether we should load Katello's org!

Hopefully this makes sense, please let me know what you think :)

waldenraines commented:
Since this is being targeted for a release that is soon and will require a cherry-pick does it make sense to put in my hack (https://github.com/Katello/katello/pull/8022) for now?  I'm concerned about the complexity of this change.

johnpmitsch commented:
@waldenraines yes, that is fine as long as we intend to continue with the refactor and remove the full page refresh

johnpmitsch commented:
Discussed this off-thread, keeping Katello's org in sync with Foreman's org inside of Katello will lead to problematic code, there is no clean way to do so we can see (others have attempted as well).

We decided that all organization management should be on the Foreman side and there should be a mechanism for plugins to load additional org info.

I'm going to close this out, if any part of this PR is helpful to solve the organization issue, feel free to use it!
